Wasm example improvements

vuldin opened the pull request:
This PR contains a few updates:
- dependencies are updated for the wasm project
- Multiple updates to three READMEs: 1) main, 2) the JS client, and 3) wasm app
- switch to avsc mainly due to details found [here](https://github.com/Adelost/javascript-serialization-benchmark)

The updated README contains steps for running the app: https://github.com/vuldin/redpanda-examples/blob/wasm-example-improvements/wasm/js/transform_avro/README.md

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda-examples?pullRequest=12) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda-examples?pullRequest=12) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda-examples?pullRequest=12) it.</sub>

vuldin commented on wasm/js/transform_avro/src/main.js:
```diff
@@ -1,4 +1,4 @@
-const avro = require("avro-js");
+const avro = require("avsc");
```
The only important change to this file is switching `avro-js` for `avsc` (see top post for link to benchmarks showing why). The rest of the changes are just name changes and could be left out. ended up including them because this branch was what I used as the basis for multiple attempts at getting both Avro and Protobuf schemas to work together in a single transform. This included reading from file, reading from registry, and various ways to include asynchronous calls in this file so that it would work after being successfully deployed.

[Here](https://github.com/vuldin/redpanda-examples/blob/two-transforms-fail/wasm/js/transform_avro/src/main.js) is one of those attempts where I'm using only synchronous code to successfully encode JSON to both Avro and Protobuf. The problem with this attempt is later where I'm adding two entries to the Map object that gets returned from `processRecord`. More details on the issue [here](https://github.com/redpanda-data/redpanda/issues/4090).

The renaming in this file would have made (and will make) way more sense with Avro and Protobuf transforms.

graphcareful replied:
Agree with the renaming +1

As for the issue you observed, I'm pretty sure I fixed that bug and the change landed in the `wasm-si-beta` release. When you initially ran into that issue did you use the old version that was listed in the `package.json` file in this project?

vuldin replied:
> Agree with the renaming +1
> 
> As for the issue you observed, I'm pretty sure I fixed that bug and the change landed in the `wasm-si-beta` release. When you initially ran into that issue did you use the old version that was listed in the `package.json` file in this project?

Yep that was the issue! I posted more details [here](https://github.com/redpanda-data/redpanda/issues/4090).

Thanks for the review, I'll update the PR later today.

vuldin commented:
This PR resolves an error when attempting to run `npm run build` with node@17:
```node:internal/crypto/hash:67
  this[kHandle] = new _Hash(algorithm, xofLen);
                  ^

Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:67:19)
    at Object.createHash (node:crypto:135:10)
    at module.exports (/Users/josh/projects/redpanda/redpanda-examples/wasm/js/transform_avro/node_modules/webpack/lib/util/createHash.js:135:53)
    at NormalModule._initBuildHash (/Users/josh/projects/redpanda/redpanda-examples/wasm/js/transform_avro/node_modules/webpack/lib/NormalModule.js:417:16)
    at handleParseError (/Users/josh/projects/redpanda/redpanda-examples/wasm/js/transform_avro/node_modules/webpack/lib/NormalModule.js:471:10)
    at /Users/josh/projects/redpanda/redpanda-examples/wasm/js/transform_avro/node_modules/webpack/lib/NormalModule.js:503:5
    at /Users/josh/projects/redpanda/redpanda-examples/wasm/js/transform_avro/node_modules/webpack/lib/NormalModule.js:358:12
    at /Users/josh/projects/redpanda/redpanda-examples/wasm/js/transform_avro/node_modules/loader-runner/lib/LoaderRunner.js:373:3
    at iterateNormalLoaders (/Users/josh/projects/redpanda/redpanda-examples/wasm/js/transform_avro/node_modules/loader-runner/lib/LoaderRunner.js:214:10)
    at Array.<anonymous> (/Users/josh/projects/redpanda/redpanda-examples/wasm/js/transform_avro/node_modules/loader-runner/lib/LoaderRunner.js:205:4)
    at Storage.finished (/Users/josh/projects/redpanda/redpanda-examples/wasm/js/transform_avro/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:55:16)
    at /Users/josh/projects/redpanda/redpanda-examples/wasm/js/transform_avro/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:91:9
    at /Users/josh/projects/redpanda/redpanda-examples/wasm/js/transform_avro/node_modules/graceful-fs/graceful-fs.js:123:16
    at FSReqCallback.readFileAfterClose [as oncomplete] (node:internal/fs/read_file_context:68:3) {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}
```

vuldin commented:
I pushed an update to the `switch avro-js for avsc` commit, which changed the output topic name to match what was referenced in the README (`avro`).

vuldin commented:
Pushed an update that makes this PR compatible with another PR in review https://github.com/redpanda-data/redpanda-examples/pull/18. Once reviewed/approved, either PR can now be merged first.

graphcareful commented on wasm/js/transform_avro/package.json:
```diff
@@ -12,12 +12,12 @@
   "license": "ISC",
   "dependencies": {
-    "@vectorizedio/wasm-api": ">=21.11.3-si-beta5",
+    "@vectorizedio/wasm-api": ">=22.1.1-rc1",
```
I think you should be using the wasm-beta release [v21.12.1-wasm-beta1](https://github.com/redpanda-data/redpanda/releases/tag/v21.12.1-wasm-beta1)

Versions older then this are most likely susceptible to observing that header check exception that you observed

graphcareful commented on wasm/js/transform_avro/README.md:
````diff
@@ -27,44 +41,68 @@ npm test       # run mocha tests
 
 ```bash
-rpk wasm deploy dist/main.js --name json2avro --description "Transforms JSON to AVRO" --brokers localhost:19092
+> rpk wasm deploy dist/main.js --name json2avro --description "Transforms JSON to AVRO" --brokers localhost:19092
````
nit: Maybe for the docs keeping the kafka port as `9092` would lend to less possibilities of confusion

graphcareful reviewed (changes requested):
Really nice work, thanks for the contribution! Only non nit comment from me was to keep the version consistent with the technical preview release.

vuldin commented:
Added a new commit that addressed feedback (thanks @graphcareful !). I also realized that `upload-schema.js` was included in ESM module format. This was done because [another PR](https://github.com/redpanda-data/redpanda-examples/pull/18) adds a dependency that only supports ESM modules and `producer.js` is refactored there. But since that PR isn't merged I'll wait to add `upload-schemas.js` until then.

vuldin commented:
Let me know if anyone thinks I should squash the commits and modify them so they make more sense based on what is changing in this PR.

jrkinley commented on clients/js/README.md:
````diff
@@ -0,0 +1,28 @@
+## Start Redpanda
+
+```bash
+cd redpanda-examples/wasm
````
I think we should move all of the compose examples to [./docker-compose](https://github.com/redpanda-data/redpanda-examples/tree/main/docker-compose).

vuldin replied:
> I think we should move all of the compose examples to [./docker-compose](https://github.com/redpanda-data/redpanda-examples/tree/main/docker-compose).

This makes sense. I'll move the compose and related redpanda config to the `docker-compose` folder. I'll also get the latest commits from main branch, and then ensure this PR's updates take into account those changes (especially regarding the new producer updates).

jrkinley reviewed (changes requested):
Thanks for this @vuldin. My only comment is I think we should consolidate all Docker Compose examples in the same top-level directory. Happy to do it in this PR or a follow up PR.

vuldin commented:
I've updated the branch with the following changes:
- move compose.yaml and redpanda.yaml into docker-compose directory (and rename files)
- rebase from main branch (handle merge conflicts)
- update docs to reflect updates from previous commits to main
- add cleanup section to wasm README
- update js client README to reflect docker compose file changes

vuldin commented:
I don't like the commit history... I would like to have fixed that prior to a PR. But this PR has been around a while and review has already started, so I'm ok with keeping it as-is if everyone else is.
